exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java

```diff
@@ -387,62 +434,102 @@ private InfoSnapshot makeTargetInfo(Resource resource) {
    *
+   * @param resource optional resource (attributes) to be converted.
    * @param scope will be converted to {@code otel_scope_*} labels if {@code otelScopeEnabled} is
    *     {@code true}.
    * @param attributes the attributes to be converted.
    * @param additionalAttributes optional list of key/value pairs, may be empty.
    */
+  @SuppressWarnings({"rawtypes", "unchecked"})
   private Labels convertAttributes(
+      @Nullable Resource resource,
       @Nullable InstrumentationScopeInfo scope,
       Attributes attributes,
       String... additionalAttributes) {
-    int numberOfScopeAttributes = 0;
-    if (otelScopeEnabled && scope != null) {
-      numberOfScopeAttributes = scope.getVersion() == null ? 1 : 2;
-    }
-    String[] names =
-        new String[attributes.size() + numberOfScopeAttributes + additionalAttributes.length / 2];
-    String[] values = new String[names.length];
-    int[] pos = new int[] {0}; // using an array because we want to increment in a forEach() lambda.
+
+    List<AttributeKey<?>> allowedAttributeKeys =
+        addResourceAttributesAsLabels
+            ? filterAllowedResourceAttributeKeys(resource)
+            : Collections.emptyList();
+
+    Map<String, String> labelNameToValue = new HashMap<>();
     attributes.forEach(
-        (key, value) -> {
-          names[pos[0]] = sanitizeLabelName(key.getKey());
-          values[pos[0]] = value.toString();
-          pos[0]++;
-        });
+        (key, value) -> labelNameToValue.put(sanitizeLabelName(key.getKey()), value.toString()));
+
     for (int i = 0; i < additionalAttributes.length; i += 2) {
-      names[pos[0]] = additionalAttributes[i];
-      values[pos[0]] = additionalAttributes[i + 1];
-      pos[0]++;
+      labelNameToValue.putIfAbsent(
+          requireNonNull(additionalAttributes[i]), additionalAttributes[i + 1]);
     }
+
     if (otelScopeEnabled && scope != null) {
-      names[pos[0]] = OTEL_SCOPE_NAME;
-      values[pos[0]] = scope.getName();
-      pos[0]++;
+      labelNameToValue.putIfAbsent(OTEL_SCOPE_NAME, scope.getName());
       if (scope.getVersion() != null) {
-        names[pos[0]] = OTEL_SCOPE_VERSION;
-        values[pos[0]] = scope.getVersion();
-        pos[0]++;
+        labelNameToValue.putIfAbsent(OTEL_SCOPE_VERSION, scope.getVersion());
       }
     }
+
+    if (resource != null) {
+      Attributes resourceAttributes = resource.getAttributes();
+      for (AttributeKey attributeKey : allowedAttributeKeys) {
+        Object attributeValue = resourceAttributes.get(attributeKey);
+        if (attributeValue != null) {
+          labelNameToValue.putIfAbsent(
+              sanitizeLabelName(attributeKey.getKey()), attributeValue.toString());
+        }
+      }
+    }
+
+    ArrayList<String> names = new ArrayList<>();
+    ArrayList<String> values = new ArrayList<>();
+    labelNameToValue.forEach(
+        (key, value) -> {
+          names.add(key);
+          values.add(value);
+        });
```

Comment:
```suggestion
    String[] names = new String[labelNameToValue.size()];
    String[] values = new String[labelNameToValue.size()];
    int[] pos = new int[]{0};
    labelNameToValue.forEach((key, value) -> {
      names[pos[0]] = key;
      names[pos[0]] = value;
      pos[0]
```

Reply:
Good suggestions. I modified it a bit, since I prefer to avoid `int[]` to keep only one variable.

Reply:
The `int[]` allows us to use a .forEach` which avoids the pesky allocations associated with `Map.Entry` (remember those!). I know we're committed to minimizing allocations in OTLP and have gone in a different direction with prometheus in using the prometheus client library which makes it impossible be as efficient with allocations, but I still think that when its possible and easy, we should still do the little things to reduce resource usage.

Reply:
No problem. Changed it.